packages/core/src/interpreter.ts

```diff
@@ -668,9 +668,28 @@ export class Interpreter<
             // TODO: determine how this can be immutably updated
             this.state.children[id] = ref;
 
+            let latestData;
+
             ref.subscribe({
               next: (data) => {
-                this.send(data as TODO);
+                latestData = data;
+                if (
+                  typeof data === 'object' &&
+                  data !== null &&
+                  'type' in data
+                ) {
+                  this.send(data as TODO);
+                } else {
+                  this.send(
+                    toSCXMLEvent(
+                      {
+                        type: `xstate.emit.${id}`,
```

Comment:
Could we say that this replaces `.sync` + `xstate.update`? It seems that with this change they are almost the same - while the `xstate.emit.*` is not sent conditionally so the parent will get "slammed" with events (I wouldn't bother optimizing this right now, just mentioning this fact as this will produce a lot of noisy events in the event log)

Reply:
on top of that - the emitted value is really unanimous to a snapshot, could we rename "emit" to snapshot in all the related things?

I know that `getSnapshot` might sometimes return `T | undefined` while the emitted value is always `T` but that seems to be a really small difference that might be handled differently without having more names for almost the same thing

Reply:
A snapshot is the latest emitted value:

1. emitted
2. emitted
3. emitted
4. emitted ← snapshot (latest)

The difference is that the snapshot is _taken_ and the emitted value is _given_, so it might be more confusing to combine these two concepts.

Reply:
Re: noisiness, we can do some optimizations in the future:

- Short-circuiting unknown events if there are no enabled wildcard transitions and completely ignoring them in the machine (no state transition occurs)
- Not subscribing to an actor if no `onEmit` nor `xstate.emit.*` transition is defined for that actor (if a tree falls in the forest...)

Reply:
> The difference is that the snapshot is taken and the emitted value is given, so it might be more confusing to combine these two concepts.

It's still the same concept/type though - it's a standard practice to have `on*` named after a thing that is being "emitted". If somebody needs to be more precise about the timing and the context of the usage they can always say something like "emitted snapshot", while `getSnapshot` just returns "last emitted snapshot".

IMHO calling those differently makes for a harder teach. WDYT @mattpocock ?

Reply:
If we name this transition `onSnapshot: { ... }`, it doesn't make sense, because it's not a snapshot; it's what the actor is emitting. A snapshot is something done by the reader, not the writer.

Reply:
> it's not a snapshot; it's what the actor is emitting. A snapshot is something done by the reader, not the writer.

I wouldn't describe it this way - mainly because the reader doesn't really manage it, it just gets it from the writer that maintains it internally. Otherwise `getSnapshot` would be conceptually an async request. It sounds to me that it's exactly the same as with BehaviorSubjects - they emit values and u can always just read it synchronously using `.getValue`. That is allowed by the fact that the behavior subject itself (and not its consumer) is responsible for managing this. 

I would agree with you IF what an actor is emitting wouldn't always be available as its snapshot - but it is.

Reply:
With XState it's inverted - the _reader_ (which creates the `ActorRef`) maintains the snapshot, not the writer. 

Conceptually, it should be thought of as this:

```js
// pseudocode
function parent() {
  const actorRef = spawnSomeActor();
  let snapshot;

  actorRef.subscribe({
    next: (emitted) => { snapshot = emitted; }
  });

  actorRef.getSnapshot = () => snapshot;

  // ...
}
```

The `ActorRef` is owned by the reader; it's the reader's "proxy" for communicating with a possibly remote actor that it doesn't own. That's why `.getSnapshot()` is always sync.

Reply:
I agree with everything said in your comment - I just don't strongly see the connection between this and the chosen naming. 

I understand that an actor can emit just anything, and in theory, that value can be adapted on the reader's side but that just complicates the things for a very little gain. Especially since this just doesn't happen in our own implementation. 

As long as the `ActorRef`'s declaration looks smth like this:
```ts
interface ActorRef<TEvent extends { type: string }, TEmitted> {
  // ...
  subscribe: (emitted: TEmitted) => (() => void);
  getSnapshot: () => TEmitted
}
```
There is a very, very strong relationship between those things. The same type appears at both of those positions. We are often also thinking about emitting events in the frontend, so using `TEmitted` might make think somebody that this is supposed to be an event, but there is no such restriction. So I think that "snapshot" makes more sense because it's more commonly associated with some kind of data/state (and that really is what we are dealing with here).

Reply:
I feel that it's especially important doubt because there were requests to allow listening for events coming from a specific actor and I think that `onEmit` could be easily mistaken for that.